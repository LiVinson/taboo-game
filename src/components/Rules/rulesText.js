
const rules = [
    {
        title: "Setup",
        text: ["There are 2 teams per game. Each round, a giver is chosen from one team, and a watcher is chosen from the other team.</p><> The giver and watcher will rotate between teams after each round."]

    }, {
        title: "Giver",
        text: ["Each round, the giver will see a card. The first word is the one you want your team to guess. The other words are words you cannot say! No gestures, rhyming word, ‘sounds like’ clues allowed", "Select Next if they guess correctly, or skip if you give up and want to display the next card. Keep going as fast as you can until time runs out!"]
    },{
        title: "Watcher",
        text: ["Each round, a watcher is chosen from the team that is not giving the clues. The watcher will be able to see the same cards as the giver, and is in charge of making sure the giver does not say any taboo words or give clues that break the rules.", "At the end of the round, all of the cards will be displayed and the watcher will confirm the status of cards: correct, skipped, or no status (cards where the giver said an illegal word or did not get their team to guess."]
    },
    {
        title: "Scoring",
        text: ["The giver’s team receives one point for each correct guess. The watcher’s team receives up to one point (depending on rules selected) for each card the giver skips. If the giver gives an illegal clue or their team cannot guess the word before time is up, no points are given to either team for that card.","The team with the most points at the end of the game wins!"]
    },
]

export default rules