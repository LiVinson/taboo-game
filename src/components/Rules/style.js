import styled from 'styled-components'

export const RulesTitle = styled.h5`
	font-family: ${(props) => props.theme.text.display};
	font-size: 2rem;
	font-weight: 500;
	text-align: center;
	color: ${(props) => props.theme.color.grayDark2};

	/* Faint gray line after section title*/
	&::after {
		display: block;
		content: '';
		width: 80%;
		height: 1.5px;
		background-color: ${(props) => props.theme.color.lightGray2};
		margin: 0.8rem auto 1rem auto;
	}
`

export const RulesText = styled.p`
	font-size: 1.6rem;
	text-align: left;
	padding: 0 0.5rem;
	margin-bottom: 1rem;
	color: ${(props) => props.theme.color.grayDark2};
`
