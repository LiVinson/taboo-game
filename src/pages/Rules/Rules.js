import React from "react"
import TabooCard from "../../components/TabooCard"
import { Redirect } from "react-router-dom"
import rules from "./rulesText"

export default function Rules({ match }) {
  const { topic } = match.params
  console.log(topic)
  return (
    <TabooCard
         tabooWord="How to Play" list={null}
         buttons={true}
    >
      <RulesText topic={topic} />
    </TabooCard>
  )
}

function RulesText({topic}) {
    console.log(topic)

    const rulesInfo = rules[parseInt(topic)]

    if (rulesInfo) {
        return (
            <React.Fragment>
                <h5>{rulesInfo.title}</h5>
                {rulesInfo.text.map((paragraph, index) => <p key={index}>{paragraph}</p>)}
            </React.Fragment>
        )
        
    } else {
        return <Redirect to="/404" />

    }
}


