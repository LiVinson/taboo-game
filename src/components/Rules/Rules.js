import React from 'react'
import { Redirect } from 'react-router-dom'
import PropTypes from 'prop-types'
import { ButtonTabooCard } from 'components/shared/TabooCard'
import rules from './rulesText'
import { RulesTitle, RulesText } from './style'

/*
  Can be displayed in 2 ways: on it's own route or within the play route.
  If on home route closing instructions goes back to /home. If on play route,
  calls a callback provided from GameInfo that will toggle Rules closed.
*/
export default class Rules extends React.Component {
	constructor(props) {
		super(props)
		this.state = {
			currentRule: 0,
		}
	}

	//When user clicks Back. If already showing first rule, either call toggleGameInfo (during game play)
	//or return to home route. Otherwise set state to previous rule
	displayPreviousRule = (event) => {
		const { currentRule } = this.state
		if (currentRule > 0) {
			this.setState((state) => {
				return {
					currentRule: state.currentRule - 1,
				}
			})
		} else {
			if (this.props.toggleGameInfo) {
				this.props.toggleGameInfo(event)
			} else {
				this.props.history.push('/home')
			}
		}
	}

	//When user clicks Next. If already showing last rule, either call toggleGameInfo (during game play)
	//or return to home route. Otherwise set state to next rule
	displayNextRule = (event) => {
		const { currentRule } = this.state
		if (currentRule < rules.length - 1) {
			this.setState((state) => {
				return {
					currentRule: state.currentRule + 1,
				}
			})
		} else {
			if (this.props.toggleGameInfo) {
				this.props.toggleGameInfo(event)
			} else {
				this.props.history.push('/home')
			}
		}
	}

	render() {
		const { currentRule } = this.state
		const buttonInfo = [
			{
				text: currentRule > 0 ? 'Back' : 'Close',
				onClick: this.displayPreviousRule,
				name: this.props.toggleGameInfo && 'showRules',
				type: 'button',
			},
			{
				text: currentRule < rules.length - 1 ? 'Next' : 'Close',
				onClick: this.displayNextRule,
				name: this.props.toggleGameInfo && 'showRules',
				type: 'button',
			},
		]

		return (
			<ButtonTabooCard tabooWord="How to Play" buttons={buttonInfo} type="home">
				<DisplayRulesText topicId={currentRule} />
			</ButtonTabooCard>
		)
	}
}

Rules.propTypes = {
	history: PropTypes.object,
	toggleGameInfo: PropTypes.func,
}

//Loops over imported array of rules and displays title and paragraphs for rule displayed
const DisplayRulesText = ({ topicId }) => {
	const rulesInfo = rules[topicId]
	if (rulesInfo) {
		return (
			<React.Fragment>
				<RulesTitle>{rulesInfo.title}</RulesTitle>
				{rulesInfo.text.map((paragraph, index) => (
					<RulesText key={index}>{paragraph}</RulesText>
				))}
			</React.Fragment>
		)
	} else {
		//Shouldn't be possible, but just in case.
		return <Redirect to="/404" />
	}
}

DisplayRulesText.propTypes = {
	topicId: PropTypes.number.isRequired,
}
