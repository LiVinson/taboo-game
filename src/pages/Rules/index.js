import Rules from "./Rules"

export default Rules