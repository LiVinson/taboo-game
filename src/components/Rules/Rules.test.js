import React from "react"
import Rules from "./Rules"
import renderer from "react-test-renderer"
import { mount } from "enzyme"


it("renders correctly", ()=> {
    const props = {
        match: {
            params: {
                topic: 0
            }
        },
        history: {
            push: jest.fn()
        }
    }

    const RulesComponent = renderer.create(<Rules {...props} />).toJSON()
    expect(RulesComponent).toMatchSnapshot()
})

it("navigates to previous rule if 'back' is selected and first rule is not displayed", ()=> {
    const props = {
        match: {
            params: {
                topic: 2
            }
        },
        history: {
            push: jest.fn()
        }
    }

    const RulesComponent = mount(<Rules {...props} />)
    RulesComponent.find("button").at(0).simulate("click")
    //Check that the history method was called

    expect(props.history.push.mock.calls[0].length).toBe(1)
        //Check the argument the push method was called with (equals the new url)
    expect(props.history.push.mock.calls[0][0]).toBe("/home/rules/1")
})

it("navigates to homepage if 'back' is selected and first rule is already displayed", ()=> {   
    const props = {
        match: {
            params: {
                topic: 0
            }
        },
        history: {
            push: jest.fn()
        }
    }
    const RulesComponent = mount(<Rules {...props} />)
    RulesComponent.find("button").at(0).simulate("click")

    //Check that the history method was called
    expect(props.history.push.mock.calls[0].length).toBe(1)
    //Check the argument the push method was called with (equals the new url)
    expect(props.history.push.mock.calls[0][0]).toBe("/home")
})

it("navigates to next rule if 'next' is selected and last rule is not displayed", ()=> {   
    const props = {
        match: {
            params: {
                topic: 1
            }
        },
        history: {
            push: jest.fn()
        }
    }
    const RulesComponent = mount(<Rules {...props} />)
    RulesComponent.find("button").at(1).simulate("click")
    //Check that the history method was called

    expect(props.history.push.mock.calls[0].length).toBe(1)
    //Check the argument the function was called with
    expect(props.history.push.mock.calls[0][0]).toBe("/home/rules/2")
})

it("navigates to home if 'next' is selected and last rule is already displayed", ()=> {   
    const props = {
        match: {
            params: {
                topic: 3 
            }
        },
        history: {
            push: jest.fn()
        }
    }
    const RulesComponent = mount(<Rules {...props} />)
    RulesComponent.find("button").at(1).simulate("click")
    //Check that the history method was called

    expect(props.history.push.mock.calls[0].length).toBe(1)
    //Check the argument the function was called with
    expect(props.history.push.mock.calls[0][0]).toBe("/home")
})