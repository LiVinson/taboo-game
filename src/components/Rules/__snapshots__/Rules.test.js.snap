// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="taboo-card taboo-card--home"
>
  <div
    className="taboo-card__word-container"
  >
    <p
      className="taboo-card__word"
    >
      How to Play
    </p>
  </div>
  <div
    className="taboo-card__list-container"
  >
    <h5>
      Setup
    </h5>
    <p>
      There are 2 teams per game. Each round, a giver is chosen from one team, and a watcher is chosen from the other team.&lt;/p&gt;&lt;&gt; The giver and watcher will rotate between teams after each round.
    </p>
  </div>
  <div
    className="btnContainer"
  >
    <button
      className="button "
      onClick={[Function]}
    >
      Close
    </button>
    <button
      className="button button--primary-color"
      onClick={[Function]}
    >
      Next
    </button>
  </div>
</div>
`;
